lib/SIL/SILFunctionType.cpp

```diff
@@ -2231,8 +2235,28 @@ namespace {
         return origType;
       }
 
-      return TheSILModule.Types.getLoweredType(abstraction, substType)
-               .getSwiftRValueType();
+      CanType SubstTy;
+      if (!SubstSig)
+        SubstTy = TheSILModule.Types.getLoweredType(abstraction, substType)
+                      .getSwiftRValueType();
+      else {
+        // In case of mapping interface types to interfaces types,
+        // the lowering should happen in the target generic context.
+
+        // Temporarily replace the generic context.
+        TheSILModule.Types.popGenericContext(genericSig);
```

Comment:
This is a workaround for rdar://29711782. There is a bug in substGenericArgs. If you remove this nonsense, but fix the issue, I'm fine with it ;-)